Compile Failure (`-Werror=format-overflow`)

**Describe the bug**
On an aarch64, archlinux arm platform using `cc` (GCC) 14.2.1 20250207, following the build instructions in the Readme results in a failure at `make` (step 5). The compiler reports that up to 255 bytes are being written to a buffer of size 254. A brief exploration of the error did not immediately reveal the source location. The full build log is reproduced below. Notably, disabling `-Werror=format-overflow` results in a successful build, but `peppa` (even though the build-log then says it was built) does not exist—only `cli` (which reports that `compile` is an invalid subcommand), `libpeppa.so` and `libpeppa.pc` are added to the `build` directory.

**To Reproduce**
Steps to reproduce the behavior:
1. Follow the build instructions precisely, up through step 4
2. Run `make` (step 5)

**Expected behavior**
All targets the `README` discusses are built and functional (i.e., `peppa`), with no errors (as `-Werror` is included in the build rules).

**Log**

<details>
<summary>Full build log</summary>

```
[ 25%] Building C object CMakeFiles/peppa.dir/peppa.c.o
/home/halosghost/prj/pak/peppapeg-git/src/PeppaPEG/peppa.c: In function ‘P4_LoadGrammarResult’:
/home/halosghost/prj/pak/peppapeg-git/src/PeppaPEG/peppa.c:821:30: error: ‘%s’ directive writing up to 255 bytes into a region of size 254 [-Werror=format-overflow=]
  821 |         sprintf((r)->errmsg, (m), __VA_ARGS__); \
      |                              ^~~
/home/halosghost/prj/pak/peppapeg-git/src/PeppaPEG/peppa.c:751:9: note: in definition of macro ‘catch_err’
  751 |         __VA_ARGS__; \
      |         ^~~~~~~~~~~
/home/halosghost/prj/pak/peppapeg-git/src/PeppaPEG/peppa.c:4905:9: note: in expansion of macro ‘P4_EvalRaisef’
 4905 |         P4_EvalRaisef(result, "%s: %s.", P4_GetErrorString(err), evalres->errmsg);
      |         ^~~~~~~~~~~~~
/home/halosghost/prj/pak/peppapeg-git/src/PeppaPEG/peppa.c:821:9: note: ‘sprintf’ output 4 or more bytes (assuming 259) into a destination of size 256
  821 |         sprintf((r)->errmsg, (m), __VA_ARGS__); \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/halosghost/prj/pak/peppapeg-git/src/PeppaPEG/peppa.c:751:9: note: in definition of macro ‘catch_err’
  751 |         __VA_ARGS__; \
      |         ^~~~~~~~~~~
/home/halosghost/prj/pak/peppapeg-git/src/PeppaPEG/peppa.c:4905:9: note: in expansion of macro ‘P4_EvalRaisef’
 4905 |         P4_EvalRaisef(result, "%s: %s.", P4_GetErrorString(err), evalres->errmsg);
      |         ^~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [CMakeFiles/peppa.dir/build.make:79: CMakeFiles/peppa.dir/peppa.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:90: CMakeFiles/peppa.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

</details>

**Desktop (please complete the following information):**
 - OS: aarch64 archlinux ARM (rolling)

--- a/.github/workflows/check.yml
+++ b/.github/workflows/check.yml
@@ -48,9 +48,13 @@ jobs:
       run: brew install libunistring python3
       if: "contains( matrix.os, 'macos')"
 
+    - uses: actions/setup-python@v5
+      with:
+        python-version: '3.x'
+
     - name: Create Build Environment
       run: |
-        sudo python3 -mpip install PyYAML
+        python3 -m pip install PyYAML
         cmake -E make_directory ${{github.workspace}}/build
 
     - name: Configure CMake (CLANG)

--- a/.github/workflows/check.yml
+++ b/.github/workflows/check.yml
@@ -16,14 +16,17 @@ on:
 
 jobs:
   job:
-    name: ${{ matrix.os }}-make-check
+    name: ${{ matrix.os }}-${{ matrix.compiler }}-unistr-${{ matrix.unistr }}-make-check
     runs-on: ${{ matrix.os }}
     if: "!contains(github.event.head_commit.message, 'skip check')"
     strategy:
       fail-fast: false
       matrix:
+        # ubuntu-latest / ubuntu-24.04-arm cover x86_64 / aarch64 Linux;
+        # macos-latest runs on arm64.
         os:
           - ubuntu-latest
+          - ubuntu-24.04-arm
           - macos-latest
         compiler:
           - gcc
@@ -33,7 +36,7 @@ jobs:
           - Off
 
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v4
       with:
         submodules: true
 
@@ -73,7 +76,7 @@ jobs:
         make
         sudo make install
 
-        if [ "${{ matrix.compiler }}" = "clang" && "${{ matrix.os }}" = "ubuntu-latest" ]; then
+        if [[ "${{ matrix.compiler }}" = "clang" && "${{ matrix.os }}" == ubuntu* ]]; then
           export ASAN_OPTIONS="detect_leaks=1"
         fi
 

--- a/peppa.c
+++ b/peppa.c
@@ -818,8 +818,7 @@ P4_PRIVATE(void)         P4_DiffPosition(P4_String str, P4_Position* start, size
 # define P4_EvalRaisef(r,m,...) \
     do { \
         memset((r)->errmsg, 0, sizeof((r)->errmsg)); \
-        if (snprintf((r)->errmsg, sizeof((r)->errmsg), (m), __VA_ARGS__) < 0) \
-            (r)->errmsg[0] = '\0'; \
+        sprintf((r)->errmsg, (m), __VA_ARGS__); \
     } while (0);
 # define is_end(s) ((s)->pos >= (s)->slice.stop.pos)
 # define is_tight(e) (((e)->flag & P4_FLAG_TIGHT) != 0)
@@ -4903,7 +4902,9 @@ P4_LoadGrammarResult(P4_String rules, P4_Result* result) {
 
     /* eval grammar rule parse tree to grammar object. */
     catch_err(P4_PegEvalGrammar(rules_tok, evalres), {
-        P4_EvalRaisef(result, "%s: %s.", P4_GetErrorString(err), evalres->errmsg);
+        /* precision bounds the copied errmsg so the total output always
+         * fits result->errmsg; snprintf is unavailable under -ansi. */
+        P4_EvalRaisef(result, "%s: %.230s.", P4_GetErrorString(err), evalres->errmsg);
     });
 
 finalize:

--- a/peppa.c
+++ b/peppa.c
@@ -818,7 +818,8 @@ P4_PRIVATE(void)         P4_DiffPosition(P4_String str, P4_Position* start, size
 # define P4_EvalRaisef(r,m,...) \
     do { \
         memset((r)->errmsg, 0, sizeof((r)->errmsg)); \
-        sprintf((r)->errmsg, (m), __VA_ARGS__); \
+        if (snprintf((r)->errmsg, sizeof((r)->errmsg), (m), __VA_ARGS__) < 0) \
+            (r)->errmsg[0] = '\0'; \
     } while (0);
 # define is_end(s) ((s)->pos >= (s)->slice.stop.pos)
 # define is_tight(e) (((e)->flag & P4_FLAG_TIGHT) != 0)

--- a/peppa.c
+++ b/peppa.c
@@ -31,6 +31,12 @@
  * @see        https://github.com/soasme/PeppaPEG
 */
 
+/* glibc declares snprintf only for C99 and later; this exposes it when
+ * compiling with -ansi. Must be defined before any system header. */
+#ifndef _ISOC99_SOURCE
+# define _ISOC99_SOURCE 1
+#endif
+
 #include <string.h>
 #include <stdlib.h>
 #include <stdio.h>
@@ -818,7 +824,7 @@ P4_PRIVATE(void)         P4_DiffPosition(P4_String str, P4_Position* start, size
 # define P4_EvalRaisef(r,m,...) \
     do { \
         memset((r)->errmsg, 0, sizeof((r)->errmsg)); \
-        sprintf((r)->errmsg, (m), __VA_ARGS__); \
+        snprintf((r)->errmsg, sizeof((r)->errmsg), (m), __VA_ARGS__); \
     } while (0);
 # define is_end(s) ((s)->pos >= (s)->slice.stop.pos)
 # define is_tight(e) (((e)->flag & P4_FLAG_TIGHT) != 0)
@@ -2642,7 +2648,7 @@ finalize:
 P4_PUBLIC P4_String
 P4_Version(void) {
     static char version[15];
-    sprintf(version, "%i.%i.%i", P4_MAJOR_VERSION, P4_MINOR_VERSION, P4_PATCH_VERSION);
+    snprintf(version, sizeof(version), "%i.%i.%i", P4_MAJOR_VERSION, P4_MINOR_VERSION, P4_PATCH_VERSION);
     return version;
 }
 
@@ -3206,7 +3212,7 @@ P4_GetErrorMessage(P4_Source* source) {
         return NULL;
 
     memset(source->errmsg, 0, sizeof(source->errmsg));
-    sprintf(source->errmsg, "line %zu:%zu, expect %s",
+    snprintf(source->errmsg, sizeof(source->errmsg), "line %zu:%zu, expect %s",
             source->error.lineno, source->error.offset,
             source->error.rule ? source->error.rule->name : source->entry_name);
 
@@ -3229,7 +3235,9 @@ P4_GetErrorMessage(P4_Source* source) {
             strcat(source->errmsg, " (left recursion rule entry cannot be lifted)");
             break;
         case E_BACKREF_OUT_REACHED:
-            sprintf(source->errmsg + strlen(source->errmsg), " (\\%lu out reached)", source->error.expr->backref_index);
+            snprintf(source->errmsg + strlen(source->errmsg),
+                    sizeof(source->errmsg) - strlen(source->errmsg),
+                    " (\\%lu out reached)", source->error.expr->backref_index);
             break;
         case E_BACKREF_TO_SELF:
             strcat(source->errmsg, " (backref point to backref)");
@@ -3238,10 +3246,14 @@ P4_GetErrorMessage(P4_Source* source) {
             strcat(source->errmsg, " (insufficient repetitions)");
             break;
         case E_INSUFFICIENT_REPEAT2:
-            sprintf(source->errmsg + strlen(source->errmsg), " (at least %lu repetitions)", source->error.expr->repeat_min);
+            snprintf(source->errmsg + strlen(source->errmsg),
+                    sizeof(source->errmsg) - strlen(source->errmsg),
+                    " (at least %lu repetitions)", source->error.expr->repeat_min);
             break;
         case E_EXCESSIVE_REPEAT:
-            sprintf(source->errmsg + strlen(source->errmsg), " (at most %lu repetitions)", source->error.expr->repeat_max);
+            snprintf(source->errmsg + strlen(source->errmsg),
+                    sizeof(source->errmsg) - strlen(source->errmsg),
+                    " (at most %lu repetitions)", source->error.expr->repeat_max);
             break;
         case E_WRONG_BACKREF:
             strcat(source->errmsg, " (back reference not match)");

